### PR TITLE
Add event exclusion feature to funnels

### LINKED ISSUE #5553: Excluding events in between funnel steps
## Context
Shamelessly copied over from [Nail Funnel Ordering Section Specs](https://docs.google.com/document/d/1S7fwha7cph-79YaPADfFKHILCnPmBgk4rqaMu-gPZ3g/edit#heading=h.nwp8ynw69xae)

### Event exclusions
Part of building a funnel is being able to exclude users who performed a certain event. An example of this is users who upgraded their account in their activation funnel, or who received a promotion in their upgrade funnel. The functionality entails determining a specific event you want to exclude within two specific steps. So users, who performed that specific event within the steps will be automatically dropped from the funnel altogether. Note this means that they will be fully removed from consideration on the entire funnel.

### Exclusion filters
This is roughly the same as above (“Event exclusions”) except that instead of allowing users to exclude a particular event at a particular point in the funnel (i.e. between certain steps), we allow them to exclude users who did a certain event at any point in the entire funnel range (e.g. Last 7 days). 

- Master Ticket - #5367 
- Backend Specs - [Comment](https://github.com/PostHog/posthog/issues/5367#issuecomment-889860866)

@clarkus's mockup taken from [#5367](https://github.com/PostHog/posthog/issues/5367#issuecomment-890204318)

image

## A few questions/discussion topics

CC'ing @clarkus @paolodamico for some guidance

1. Are exclusions available for unordered steps? If so what would a valid step range look like for unordered steps?
2. It seems like this exclusion problem is two fold: event exclusions and user exclusions. I'm imagining user exclusion filters would best belong as another section in the 'Filters' card below. What are your thoughts on placement for exclusion features? Are they better grouped by exclusion in its separate card, by functionality, or something else?

Screen Shot 2021-08-11 at 2 36 59 PM

### LINKED ISSUE #5584: "There are no matching events for this query" message for a slow-loading funnel is confusing
## Bug description
When a funnel is taking a long time to load (1+ minute), it shows the following contents:

1. A spinner
2. A screen that says "There are no matching events for this query"
3. A screen that says "Looks like things are a little slow"

Screen Shot 2021-08-16 at 1 56 13 PM
Screen Shot 2021-08-16 at 2 12 22 PM

Since Step #2 can hang around a long time, it's caused confusion for many of our internal users.

## Expected behavior
Don't show step #2.

## How to reproduce
1. Create an extremely slow query in a funnel (1+ minutes to load)
2. Wait & watch

## Environment

- [ ] PostHog Cloud
- [x] self-hosted PostHog

## Additional context
Our deployment has recently seen its resources upgraded; this may not appear anymore if the deployment is sufficiently quick.

### PR DESCRIPTION
**Pull request by alexkim205**
## Changes

This PR adds the ability to exclude events between funnel steps.

Closes #5553.
Addresses part of #5367.

- [x] adding some functionality to `<ActionFilter/>`
- [x] adding new `<FunnelExclusionsFilter/>`
- [x] consolidated and standardized how we render empty states (loading, error, timeout, and insight-specific empty states like - we have for funnels) across all of our insight graphs. This refactor fixes #5584 and a few of other general insight loading bugs we've seen in the past. (see `BlockingEmptyState` and `CoexistingEmptyState` in code).
- [x] add warning empty state if exclusion filters are invalid b/c they overlap funnel steps.

Because this is a bit of a chunky PR, I'll CC a few people for focused reviews.

- @clarkus UX
- @macobo / @mariusandra code quality
- @paolodamico product/copy
- @neilkakkar interfacing with the api

Demo

https://user-images.githubusercontent.com/13460330/129821381-c52d927c-cdf5-4104-97a0-4cc72d7771f3.mov

Horizontal view

Screen Shot 2021-08-17 at 6 17 43 PM

Vertical view

https://user-images.githubusercontent.com/13460330/129975149-c8663893-dfbe-4e9a-b7fb-558e95b7dbd3.mov

Exclusion filters are invalid

https://user-images.githubusercontent.com/13460330/130532685-662a8afa-a986-4c26-bf48-0c50909b6ac1.mov

## Checklist

- [ ] All querysets/queries filter by Organization, by Team, and by User
- [ ] Django backend tests
- [ ] Jest frontend tests
- [ ] Cypress end-to-end tests
- [ ] Migrations are safe to run at scale (e.g. PostHog Cloud) – present proof if not obvious
- [ ] New/changed UI is decent on smartphones (viewport width around 360px)

### CODE REVIEW

File: frontend/src/scenes/funnels/funnelUtils.ts
Comment (macobo) [inline frontend/src/scenes/funnels/funnelUtils.ts:144]: Anything we can do to clear this up? cc @neilkakkar

Comment (neilkakkar) [reply frontend/src/scenes/funnels/funnelUtils.ts:144]: Is this because of the `postgres` based loading?

I think we did away with this for Clickhouse. Optimising this would fall under https://github.com/PostHog/posthog/issues/5313 - doing away with polling.

**Review by paolodamico — CHANGES_REQUESTED** [review]
Great stuff @alexkim205! Some feedback inline and below,

- I can select exclusions within the same step.
    
- When adding a new exclusion step, maybe we should default to Pageview too? Not sure, CC @clarkus 
- UI doesn't look great on smaller screens (13" MB), think we could fit the dropdowns in a single line and align the remove button better. 
    
- While testing things out I realized that having a hide/disable button could be quite useful as when you apply these filters you want to quickly notice the difference in numbers with and without the filter, could adding this behavior be helpful? CC @clarkus 
- Noticed some weird behavior (though not sure if this is Core Analytics scope, CC @neilkakkar). Tested using vanilla demo data. Without the exclusion step my universe is 43 users, yet when I apply the exclusions step, not only are users excluded from converting to step 2, but also they're removed from consideration since step 1.

**Without exclusion step**
    


**With exclusion step**

File: frontend/src/scenes/insights/InsightTabs/FunnelTab/FunnelSecondaryTabs.tsx
Comment (neilkakkar) [reply frontend/src/scenes/insights/InsightTabs/FunnelTab/FunnelSecondaryTabs.tsx:33]: Yeah correct, we exclude users who did specific events (the exclusion events) between the funnel steps.

To answer your earlier question: 
> Noticed some weird behavior (though not sure if this is Core Analytics scope, CC @neilkakkar). Tested using vanilla demo data. Without the exclusion step my universe is 43 users, yet when I apply the exclusions step, not only are users excluded from converting to step 2, but also they're removed from consideration since step 1.

The users are indeed removed from consideration, that's intended, as otherwise the conversion rates get messed up (counted in step 1, but forced to eject from step 2 because of exclusion)

**Comment by alexkim205** [comment]
Thanks @paolodamico! 

> I can select exclusions within the same step.

Fixed!

> * UI doesn't look great on smaller screens (13" MB), think we could fit the dropdowns in a single line and align the remove button better.

Hmm, I tested this out initially but the width of the left options space was too small to fit three dropdowns in one single line without hiding all the text in the fields. Perhaps keeping this column format, but expanding the step dropdowns to their full width would fill out the container a bit better? What do you think?

**Comment by clarkus** [comment]
> When adding a new exclusion step, maybe we should default to Pageview too? Not sure, CC @clarkus

I think exclusions should work exactly the same as defining a graph series, just with the step inputs included. Defaulting to pageview makes sense to me. 

> UI doesn't look great on smaller screens (13" MB), think we could fit the dropdowns in a single line and align the remove button better.

👍 looks like we have sufficient space to make everything inline here. That would be a worthwhile change to make. 

> While testing things out I realized that having a hide/disable button could be quite useful as when you apply these filters you want to quickly notice the difference in numbers with and without the filter, could adding this behavior be helpful? CC @clarkus

hmmm I could see this being useful but I would also question if it's more-broadly applicable. There were some ideas for toggling metrics on and off via a legend control, but this is a bit deeper in the query-building side of things. I think it would depend on how difficult it is to re-specify that step. We could have toggles for hiding or showing, but we could also provide some simple "undo / redo" actions to accomplish the same end. I will think about this more. It would also be good to validate it as something useful to users. I think it's a solid idea, just not sure it should be built into this initial version until we can trace it back to a problem.

**Review by clarkus — APPROVED** [review]
I think this is looking good from a UX standpoint. Totally agree with @paolodamico's recent round of feedback.

**Comment by alexkim205** [comment]
Thanks Chris! 

I bumped up the width of the funnel tab in #5644, so let me see if I can fit everything into one line again.

@paolodamico I'll shelve the disable/hide button for now in this PR but glad to pick it up if users start demanding it in the future.

---

V.2

Hmm even with the wider tab, it seems tricky to fit everything in. Maybe we can omit the Steps prefix from each label?


https://user-images.githubusercontent.com/13460330/129970627-892f3de8-ebfe-4be4-8c89-bbf264b042f4.mov

V.3

Tried out another layout and this feels a lot easier to look at.


https://user-images.githubusercontent.com/13460330/129975078-0b0d9666-47ec-4b02-9ff3-5611b1ed401e.mov

**Comment by alexkim205** [comment]
Addressed all feedback! Please let me know what you think

**Review by clarkus — APPROVED** [review]

**Comment by paolodamico** [comment]
Great catch on the new layout @alexkim205, that is what I actually meant by single line, only the step selectors in a single line, not the entire thing. Taking another look now.

**Review by paolodamico — APPROVED** [review]
Product-wise looks good, I only found the bug below worth fixing. Other than that I would wait for @mariusandra and/or @macobo to provide feedback on the code before merging.

**Comment by alexkim205** [comment]
> Product-wise looks good, I only found the bug below worth fixing. Other than that I would wait for @mariusandra and/or @macobo to provide feedback on the code before merging.
> 
> 
> 
> 

Thanks, the overflow scroll was intentional bc it's impossible to keep the selectors on a single line on much smaller screens. Do you think wrapping the selectors like in the first iteration would better tradeoff than a scroll overflow?

**Comment by clarkus** [comment]
> Thanks, the overflow scroll was intentional bc it's impossible to keep the selectors on a single line on much smaller screens. Do you think wrapping the selectors like in the first iteration would better tradeoff than a scroll overflow?

I think we should adjust breakpoints to wrap the containers into a single column instead of introducing the scroll bars. Also note that this is pretty isolated to funnels as it's the only insight using a multi-column layout right now. If we could align the layouts across types, we might not have to contend with this issue as much.

**Comment by alexkim205** [comment]
Got it make sense and agreed. I'm not sure what the rationale behind having the multi column layout in funnels was, but it would be a much more consistent experience if funnels also was converted into a horizontal layout. @paolodamico if you think it makes sense to make this switch, I can go ahead an remove that toggle in this PR?

**Comment by alexkim205** [comment]
Here's the newest version with some flex wrapping

https://user-images.githubusercontent.com/13460330/130115538-cd81e377-7d71-4e20-86b9-c236b995bc51.mov

**Comment by alexkim205** [comment]
Changing funnels to a horizontal layout would also entail bringing filters/breakdowns/and possibly other options into the main query box like we do for other insights, so the user doesn't have to scroll down too much.


https://user-images.githubusercontent.com/13460330/130116094-0f97ebc7-62bf-46fc-9590-2f050dc3de40.mov

**Comment by paolodamico** [comment]
Think this definitely looks a lot better! Though I think my suggestion would be:
a) Move the trash icon to the first line (the event/action selector is already wide enough and also pops open)
b) When you have to move the "and" and second step dropdown to a new line, align the second dropdown horizontally to the first dropdown, so they start at the same x coordinate, IMO it'll look cleaner that way.

**Comment by clarkus** [comment]
>  a) Move the trash icon to the first line (the event/action selector is already wide enough and also pops open)

I think we should keep it right aligned. That's somewhat of an established pattern in other repeating series we have - tables, filters, etc. If it varies position it's going to cause more confusion than the wrapping rows of elements. I'd rather we drop the arrow icon instead if we're looking for a short term way to save space. I don't think we need to wrap rows. If we drop the arrow icons and optimize for a minimum width of 320px on the container, it will fit.

Screen Shot 2021-08-19 at 12 41 26 PM

We could also benefit from some kind of row marker for each distinct exclusion step. We use numbers in the funnel steps, here we could do the same or just use some other indicator.

Screen Shot 2021-08-19 at 12 48 38 PM

Longer term, this component would really benefit from some standardization. I've been working on that problem, but I need a bit more time to finish work on some edge cases.

**Comment by paolodamico** [comment]
Definitely on the right alignment, I meant keeping it to the right but on the same horizontal line as the event/action selector. Dropping the arrow also makes a lot of sense.

**Comment by alexkim205** [comment]
Thanks for waiting! It took a while to style it like this because Action filter rows are nested deeply inside a component that's used everywhere. I added a new prop called `renderRow` in `<ActionFilter/>` that lets you construct a row however you want to, without breaking any existing functionality. Hopefully one day we'll be able to refactor today's ActionFilters into subcomponents for these different layouts but I think this is a step in the right direction for future proofing/standardizing how we use this component today. 

I also tweaked our funnel tab breakpoints so that all dropdowns can fit in a single row.

Here's how the new filter looks now:


https://user-images.githubusercontent.com/13460330/130152518-ddc82f08-b315-480d-833e-176835bebcaa.mov

File: frontend/src/scenes/funnels/funnelLogic.ts
Comment (mariusandra) [inline frontend/src/scenes/funnels/funnelLogic.ts:255]: Wondering if we could do without this reducer, and just take the `exclusionFilters` directly from the filters? For good data hygiene, we should always have one source of truth, and derive whatever we need from there. In this case deriving this from the filters via selectors. Or is there some timing/recalculation issue causing problems?

File: frontend/src/scenes/funnels/funnelLogic.ts
Comment (mariusandra) [inline frontend/src/scenes/funnels/funnelLogic.ts:695]: Kind of the same question here. Can't this directly be part of the filters?

File: frontend/src/scenes/funnels/funnelUtils.ts
Comment (mariusandra) [reply frontend/src/scenes/funnels/funnelUtils.ts:148]: +1 for doing away with polling 🤞

File: frontend/src/scenes/insights/InsightTabs/FunnelTab/FunnelExclusionsFilter.tsx
Comment (mariusandra) [inline frontend/src/scenes/insights/InsightTabs/FunnelTab/FunnelExclusionsFilter.tsx:56]: Is there some way we can avoid the local state here? Introducing local state always leads to some way things can get out of sync. If there's a way to directly derive this state what what we have in a logic, the widget will be more durable.

**Review by mariusandra — DISMISSED** [review]
Hey, I gave the code a look and played with the feature a bit. 

First, is this excluded on postgres?

Second, I think there are ways the data model can be used. Ideally we'd just store everything inside the filter without using any local state: no useState in a component and no extra reducers in a logic. There may be reasons why that's not a good idea (defaults? ux limitations?), and let me know if so.

Third, I'm not sure how it should work :D. I'm locally excluding all users that ever did a pageview between all the pageviews, but no matter what exclusion I apply, nothing changes:



I might be misunderstanding how it works, but I'd expect to see no results with something like this. CC @neilkakkar

**Comment by neilkakkar** [comment]
@mariusandra thanks for raising the functionality issue. Excluding the same event as the one in the funnel shouldn't be possible! I'll update the backend to return nothing here, but maybe the frontend should warn as well - that this operation doesn't make sense.

Specifically, when you have an exclusion on event A between step X and Y, then the funnel steps from X to Y can't be A.

Edit: Actually, no, I'm thinking I should raise a ValidationError, and you can then display the error message as you see fit?

**Comment by mariusandra** [comment]
I think there are still outstanding issues from my last review. In addition to that, now that https://github.com/PostHog/posthog/pull/5668 is in, we should improve this flow somewhat:



It's weird to get such an error during seemingly normal operations. Perhaps leaving the default event for exclusions empty will be enough? Though I'd love to also see the error message somewhere in the interface.

File: frontend/src/scenes/funnels/funnelLogic.ts
Comment (alexkim205) [reply frontend/src/scenes/funnels/funnelLogic.ts:256]: That's a really good idea.  Initially I added it because it looked like `<ActionFilter>` was specifically built to act on the `events` property of `filters`, but if I had a selector and action to read and write from the `filters` source of truth than that would make for a lot cleaner less buggier code.

File: frontend/src/scenes/funnels/funnelUtils.ts
Comment (alexkim205) [reply frontend/src/scenes/funnels/funnelUtils.ts:148]: Going through these comments one by one—I'll resolve this comment to be addressed under #5313

File: frontend/src/scenes/insights/InsightTabs/FunnelTab/FunnelExclusionsFilter.tsx
Comment (alexkim205) [reply frontend/src/scenes/insights/InsightTabs/FunnelTab/FunnelExclusionsFilter.tsx:56]: Yup we can avoid this. Just made some changes getting rid of this local state stuff.

**Comment by alexkim205** [comment]
Thanks for the review @mariusandra and @neilkakkar for implementing that error. I've updated the PR description w/ my updates!

> First, is this excluded on postgres?

Thats a good catch. This is CH exclusive feature so we should protect this behind some flag.

> It's weird to get such an error during seemingly normal operations. Perhaps leaving the default event for exclusions empty will be enough? Though I'd love to also see the error message somewhere in the interface.

Gotcha just built out the below warning state for this use case. 


https://user-images.githubusercontent.com/13460330/130532408-dcd3c502-6587-46c8-bcb3-8436237853e6.mov

Other changes made:

- [ ] consolidated exclusion filters into filters so that we have one source of truth, and did away with local state.
- [ ] consolidated and standardized how we render empty states (loading, error, timeout, and insight-specific empty states like we have for funnels) across all of our insight graphs.

File: frontend/src/scenes/insights/Insights.tsx
Comment (mariusandra) [inline frontend/src/scenes/insights/Insights.tsx:184]: I guess moving this here makes sense for insights, but in turn breaks dashboards. Here's for example something with just one step that I added to my dashboard. Somehow nothing is shown. Not sure about all the other error states, as dashboard items can also timeout, etc.

image

**Review by mariusandra — APPROVED** [review]
In general this looks good! I can't guarantee I caught all bugs during the review and testing, but did notice that now funnels on dashboards had a small regression. Something to still fix before this is ready.

File: frontend/src/scenes/insights/Insights.tsx
Comment (alexkim205) [reply frontend/src/scenes/insights/Insights.tsx:184]: Nice catch—because these states can exist in both dashboards and insights, sharing these empty states across both situations should fix this.

Comment (alexkim205) [reply frontend/src/scenes/insights/Insights.tsx:184]: Fixed!

Screen Shot 2021-08-24 at 10 30 37 AM

**Comment by alexkim205** [comment]
Got some approvals and blocked on some funnel filters work so gonna go ahead and merge this one

**Pull request merged**
